File: client/lib/api/v2/schemas/simulations.ts
/**
 * Simulations V2 API Schemas
 * Schema definitions for simulations v2 endpoints
 */

import { z } from "zod";
import {
  DepartmentMappingSchema,
  ParameterItemMappingSchema,
  ParameterMappingSchema,
  RubricMappingSchema,
  ScenarioMappingSchema,
} from "./base";

// ============================================================================
// REQUEST SCHEMAS
// ============================================================================

export const SimulationsFiltersSchema = z.object({
  profileId: z.string(),
});

export type SimulationsFilters = z.infer<typeof SimulationsFiltersSchema>;

// ============================================================================
// RESPONSE SCHEMAS
// ============================================================================

// Simulation item (normalized - IDs only)
export const SimulationItemSchema = z.object({
  simulation_id: z.string(),
  name: z.string(), // Maps to simulations.title
  description: z.string(),
  department_ids: z.array(z.string()).nullable(), // None = cross-department (all departments)
  time_limit: z.number().nullable(),
  active: z.boolean(),
  practice_simulation: z.boolean(),
  can_edit: z.boolean(),
  can_delete: z.boolean(),
  can_duplicate: z.boolean(),
  scenario_ids: z.array(z.string()),
  rubric_id: z.string(),
  num_cohorts: z.number(), // Number of cohorts using this simulation
});

export const SimulationsListResponseSchema = z.object({
  simulations: z.array(SimulationItemSchema),
  scenario_mapping: ScenarioMappingSchema,
  rubric_mapping: RubricMappingSchema,
});

export type SimulationsListResponse = z.infer<
  typeof SimulationsListResponseSchema
>;
export type SimulationItem = z.infer<typeof SimulationItemSchema>;

// ============================================================================
// DETAIL SCHEMAS
// ============================================================================

export const SimulationDetailRequestSchema = z.object({
  simulationId: z.string(),
  profileId: z.string(),
});

export type SimulationDetailRequest = z.infer<
  typeof SimulationDetailRequestSchema
>;

// Scenario in simulation with position
export const ScenarioInSimulationSchema = z.object({
  scenario_id: z.string(),
  title: z.string(),
  description: z.string(),
  active: z.boolean(),
  position: z.number(),
  parameter_item_ids: z.array(z.string()),

  // Statistics fields
  usage_count: z.number(),
  success_rate: z.number(),
  last_used: z.string().nullable(),
  can_remove: z.boolean(),
});

// Parameter item schemas
export const ParameterItemSchema = z.object({
  id: z.string(),
  parameter_id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
});

export const ParameterItemDetailSchema = z.object({
  id: z.string(),
  name: z.string(),
  description: z.string().nullable(),
  parameter_id: z.string(),
});

// Parameter mapping and parameter item mapping are imported from base.ts

export const SimulationDetailResponseSchema = z.object({
  // Basic fields
  name: z.string(), // Maps to simulations.title
  description: z.string(),
  department_ids: z.array(z.string()).nullable(),
  valid_department_ids: z.array(z.string()),
  time_limit: z.number().nullable(),
  rubric_id: z.string(),
  valid_rubric_ids: z.array(z.string()),
  scenario_ids: z.array(z.string()),
  valid_scenario_ids: z.array(z.string()),

  // Boolean parameters
  active: z.boolean(),
  practice_simulation: z.boolean(),

  // Permission flags
  can_edit: z.boolean(),
  can_duplicate: z.boolean(),
  can_delete: z.boolean(),

  // Usage status
  in_use: z.boolean(),
  cohort_count: z.number(),

  // Full scenario objects (ordered by position)
  scenarios: z.array(ScenarioInSimulationSchema),

  // Parameter data for scenario picker
  parameters: z.array(ParameterItemSchema),
  parameter_items: z.array(ParameterItemDetailSchema),
  parameter_mapping: ParameterMappingSchema,

  // Top-level mappings
  scenario_mapping: ScenarioMappingSchema,
  rubric_mapping: RubricMappingSchema,
  department_mapping: DepartmentMappingSchema,
  parameter_item_mapping: ParameterItemMappingSchema,
});

export type SimulationDetailResponse = z.infer<
  typeof SimulationDetailResponseSchema
>;

// Default detail request
export const SimulationDetailDefaultRequestSchema = z.object({
  profileId: z.string(),
});

export type SimulationDetailDefaultRequest = z.infer<
  typeof SimulationDetailDefaultRequestSchema
>;

// ============================================================================
// MUTATION SCHEMAS
// ============================================================================

// Scenario in request with active state
export const ScenarioInRequestSchema = z.object({
  scenario_id: z.string(),
  active: z.boolean(),
});

export type ScenarioInRequest = z.infer<typeof ScenarioInRequestSchema>;

// Create request
export const CreateSimulationRequestSchema = z.object({
  title: z.string(),
  description: z.string(),
  department_ids: z.array(z.string()).nullable(),
  active: z.boolean(),
  practice_simulation: z.boolean(),
  time_limit: z.number().nullable(),
  rubric_id: z.string(),
  scenario_ids: z.union([
    z.array(z.string()),
    z.array(ScenarioInRequestSchema),
  ]),
});

export type CreateSimulationRequest = z.infer<
  typeof CreateSimulationRequestSchema
>;

export const CreateSimulationResponseSchema = z.object({
  success: z.boolean(),
  simulationId: z.string(),
  message: z.string(),
});

export type CreateSimulationResponse = z.infer<
  typeof CreateSimulationResponseSchema
>;

// Update request
export const UpdateSimulationRequestSchema = z.object({
  simulationId: z.string(),
  title: z.string(),
  description: z.string(),
  department_ids: z.array(z.string()).nullable(),
  active: z.boolean(),
  practice_simulation: z.boolean(),
  time_limit: z.number().nullable(),
  rubric_id: z.string(),
  scenario_ids: z.union([
    z.array(z.string()),
    z.array(ScenarioInRequestSchema),
  ]),
});

export type UpdateSimulationRequest = z.infer<
  typeof UpdateSimulationRequestSchema
>;

export const UpdateSimulationResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type UpdateSimulationResponse = z.infer<
  typeof UpdateSimulationResponseSchema
>;

// Duplicate request
export const DuplicateSimulationRequestSchema = z.object({
  simulationId: z.string(),
});

export type DuplicateSimulationRequest = z.infer<
  typeof DuplicateSimulationRequestSchema
>;

export const DuplicateSimulationResponseSchema = z.object({
  success: z.boolean(),
  simulationId: z.string(),
  message: z.string(),
});

export type DuplicateSimulationResponse = z.infer<
  typeof DuplicateSimulationResponseSchema
>;

// Delete request
export const DeleteSimulationRequestSchema = z.object({
  simulationId: z.string(),
});

export type DeleteSimulationRequest = z.infer<
  typeof DeleteSimulationRequestSchema
>;

export const DeleteSimulationResponseSchema = z.object({
  success: z.boolean(),
  message: z.string(),
});

export type DeleteSimulationResponse = z.infer<
  typeof DeleteSimulationResponseSchema
>;

File: server/app/schemas/simulations.py
"""Simulations V2 API schemas."""

from pydantic import BaseModel

from .base import (DepartmentMapping, ParameterItemMapping, ParameterMapping,
                   RubricMapping, ScenarioMapping)

# ============================================================================
# REQUEST SCHEMAS
# ============================================================================


class SimulationsFilters(BaseModel):
    """Filters for simulations list."""

    profileId: str


# ============================================================================
# RESPONSE SCHEMAS
# ============================================================================


class SimulationItem(BaseModel):
    """Simulation item in list response."""

    simulation_id: str
    name: str  # Maps to simulations.title
    description: str
    department_ids: list[str] | None  # None = cross-department (all departments)
    time_limit: int | None
    active: bool
    practice_simulation: bool
    can_edit: bool
    can_delete: bool
    can_duplicate: bool
    scenario_ids: list[str]
    rubric_id: str
    num_cohorts: int  # Number of cohorts using this simulation


class SimulationsListResponse(BaseModel):
    """Response for simulations list endpoint."""

    simulations: list[SimulationItem]
    scenario_mapping: ScenarioMapping
    rubric_mapping: RubricMapping
    department_mapping: DepartmentMapping


# ============================================================================
# DETAIL SCHEMAS
# ============================================================================


class ScenarioInSimulation(BaseModel):
    """Scenario with position in simulation."""

    scenario_id: str
    title: str
    description: str
    active: bool
    position: int  # From simulation_scenarios junction table
    parameter_item_ids: list[str]  # For displaying badges

    # Statistics fields
    usage_count: int  # Number of all chats (regardless of completion)
    success_rate: int  # Percentage (0-100) of completed chats that passed
    last_used: str | None  # ISO timestamp or None
    can_remove: bool  # True if usage_count == 0


class ParameterItem(BaseModel):
    """Parameter data for dropdown."""

    id: str
    parameter_id: str
    name: str
    description: str | None


class ParameterItemDetail(BaseModel):
    """Full parameter item details."""

    id: str
    name: str
    description: str | None
    parameter_id: str


class SimulationDetailRequest(BaseModel):
    """Request for simulation detail."""

    simulationId: str
    profileId: str


class SimulationDetailResponse(BaseModel):
    """Response for simulation detail endpoint."""

    # Basic fields
    name: str  # Maps to simulations.title
    description: str
    department_ids: list[str] | None  # None = cross-department (all departments)
    valid_department_ids: list[str]
    time_limit: int | None
    rubric_id: str
    valid_rubric_ids: list[str]
    scenario_ids: list[str]
    valid_scenario_ids: list[str]

    # Boolean parameters
    active: bool
    practice_simulation: bool

    # Permission flags
    can_edit: bool
    can_duplicate: bool
    can_delete: bool

    # Usage status
    in_use: bool
    cohort_count: int

    # Full scenario objects (ordered by position)
    scenarios: list[ScenarioInSimulation]

    # Parameter data for scenario picker
    parameters: list[ParameterItem]
    parameter_items: list[ParameterItemDetail]
    parameter_mapping: ParameterMapping

    # Top-level mappings
    scenario_mapping: ScenarioMapping
    rubric_mapping: RubricMapping
    department_mapping: DepartmentMapping
    parameter_item_mapping: ParameterItemMapping


class SimulationDetailDefaultRequest(BaseModel):
    """Request for default simulation detail."""

    profileId: str


# ============================================================================
# MUTATION SCHEMAS
# ============================================================================


class ScenarioInRequest(BaseModel):
    """Scenario with active state for create/update requests."""

    scenario_id: str
    active: bool = True


class CreateSimulationRequest(BaseModel):
    """Request to create simulation."""

    title: str
    description: str
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    practice_simulation: bool
    time_limit: int | None
    rubric_id: str
    scenario_ids: (
        list[str] | list[ScenarioInRequest]
    )  # Support both formats for backward compatibility


class CreateSimulationResponse(BaseModel):
    """Response from create simulation."""

    success: bool
    simulationId: str
    message: str


class UpdateSimulationRequest(BaseModel):
    """Request to update simulation."""

    simulationId: str
    title: str
    description: str
    department_ids: list[str] | None  # None = cross-department (superadmin only)
    active: bool
    practice_simulation: bool
    time_limit: int | None
    rubric_id: str
    scenario_ids: (
        list[str] | list[ScenarioInRequest]
    )  # Support both formats for backward compatibility


class UpdateSimulationResponse(BaseModel):
    """Response from update simulation."""

    success: bool
    message: str


class DuplicateSimulationRequest(BaseModel):
    """Request to duplicate simulation."""

    simulationId: str


class DuplicateSimulationResponse(BaseModel):
    """Response from duplicate simulation."""

    success: bool
    simulationId: str
    message: str


class DeleteSimulationRequest(BaseModel):
    """Request to delete simulation."""

    simulationId: str


class DeleteSimulationResponse(BaseModel):
    """Response from delete simulation."""

    success: bool
    message: str
